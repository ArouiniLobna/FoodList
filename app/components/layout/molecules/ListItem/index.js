/**
 *
 * ListItem
 *
 */

import React, { memo } from 'react';
import PropTypes from 'prop-types';
import CheckBox from 'components/layout/atoms/CheckBox';
import styled from 'styled-components';

function ListItem({ item, setCheckedStatus, itemKey }) {
  const handleSetCheckedStatus = (propertyName, isChecked) => {
    setCheckedStatus(itemKey, propertyName, isChecked);
  };
  return (
    <SListItem
      role="group" // as part of accessibilty we need to group the checkboxes
      aria-labelledby={`label-${itemKey}`}
      isHealthy={item.isHealthy}
      isDelicious={item.isDelicious}
    >
      <td>
        <h5 id={`label-${itemKey}`}>{item.label}</h5>
      </td>
      <td>
        <CheckBox
          checked={item.isDelicious}
          label="is Delicious"
          onChange={handleSetCheckedStatus}
          dataName="isDelicious"
        />
      </td>
      <td>
        <CheckBox
          checked={item.isHealthy}
          label="is Healthy"
          onChange={handleSetCheckedStatus}
          dataName="isHealthy"
        />
      </td>
    </SListItem>
  );
}

const SListItem = styled.tr`
  padding: 10px;
  display: flex;
  align-items: center;
  background-color: ${props => (props.isHealthy ? '#E0EEE0' : 'transparent')};
  transition: 1s;
  h5 {
    margin: 0;
    transition: 1s;
    text-indent: ${props => (props.isDelicious ? '30px' : '0')};
  }
`;

ListItem.propTypes = {
  setCheckedStatus: PropTypes.func.isRequired,
  item: PropTypes.object.isRequired,
  itemKey: PropTypes.string.isRequired, // this the food item unique identifier
};

export default memo(ListItem);
