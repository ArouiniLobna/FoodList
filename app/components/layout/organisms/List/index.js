/**
 *
 * List
 *
 */

import React, { memo } from 'react';
import PropTypes from 'prop-types';
import ListItem from 'components/layout/molecules/ListItem';
import styled from 'styled-components';

function List({ list, handleSelectionChange }) {
  // we need to know which are we updating: using the key code used within the data; Assuming it is a unique key
  // we need to know which propery got changes : isDelicious Vs isHealthy
  // we need to know if checked or unchecked
  const handleSetCheckedStatus = (identifierKey, propertyName, isChecked) => {
    handleSelectionChange(identifierKey, propertyName, isChecked);
  };

  return (
    <SList>
      <table>
        <thead>
          <tr>
            <th>Name</th>
            <th>is Delicious </th>
            <th>is Healthy</th>
          </tr>
        </thead>
        <tbody>
          {list.map(function(item) {
            return (
              <ListItem
                setCheckedStatus={handleSetCheckedStatus}
                key={Object.keys(item)[0]}
                item={item[Object.keys(item)[0]]}
                itemKey={Object.keys(item)[0]}
              />
            );
          })}
        </tbody>
      </table>
    </SList>
  );
}

const SList = styled.div`
  max-width: 400px;
  margin: 50px auto 20px;
  table {
    display: block;
    filter: drop-shadow(0px 3px 6px rgba(75, 81, 91, 0.15)),
      drop-shadow(0px 1px 3px rgba(0, 0, 0, 0.15));
    thead,
    tbody {
      display: block;
      tr {
        display: block;
        margin: 0;
      }
      th,
      td {
        width: 33.33%;
        display: inline-block;
      }
      th:not(:first-child),
      td:not(:first-child) {
        align-items: center;
        justify-content: center;
        text-align: center;
      }
    }
    thead {
      background-color: #f4f7fc;
      padding: 15px 0;
      color: #606f89;
      font-size: 12px;
      letter-spacing: 0.05em;
      text-transform: uppercase;
      font-weight: 100;
      border-bottom: 10px solid #fff;
    }
    tbody {
      background-color: #f8f8f8;
      color: #2e3b52;
      tr {
        border-bottom: 5px solid #fff;
      }
    }
  }
  * {
    box-sizing: border-box;
  }
`;
List.propTypes = {
  list: PropTypes.array,
  handleSelectionChange: PropTypes.func,
};

export default memo(List);
