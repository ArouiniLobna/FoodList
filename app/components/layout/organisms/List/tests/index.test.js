/**
 *
 * Tests for List
 *
 * @see https://github.com/react-boilerplate/react-boilerplate/tree/master/docs/testing
 *
 */

import React from 'react';
import { render } from 'react-testing-library';
import { mount } from 'enzyme';
import ListItem from 'components/layout/molecules/ListItem';
import List from '../index';

describe('<List />', () => {
  let container;
  let props;

  beforeEach(() => {
    props = {
      list: [
        {
          e5d9d9f5: {
            label: 'ice cream',
            isDelicious: true,
            isHealthy: false,
          },
        },
        {
          a9ba692b: {
            label: 'pizza',
            isDelicious: true,
            isHealthy: false,
          },
        },
        {
          ze128a47: {
            label: 'spinach',
            isDelicious: false,
            isHealthy: false,
          },
        },
      ],
      handleSelectionChange: jest.fn(),
    };
    container = mount(<List {...props} />);
  });
  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    render(<List {...props} />);
    expect(spy).not.toHaveBeenCalled();
  });

  it('expect to render a table', () => {
    expect(container.find('table').length).toBe(1);
  });

  it('expect to render a thead', () => {
    expect(container.find('thead').length).toBe(1);
  });

  it('expect to render a tbody', () => {
    expect(container.find('tbody').length).toBe(1);
  });

  it('expect to render number of ListItem as data length', () => {
    expect(container.find(ListItem).length).toBe(props.list.length);
  });

  /**
   * Unskip this test to use it
   *
   * @see {@link https://jestjs.io/docs/en/api#testskipname-fn}
   */
  it('Should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(<List {...props} />);
    expect(firstChild).toMatchSnapshot();
  });
});
