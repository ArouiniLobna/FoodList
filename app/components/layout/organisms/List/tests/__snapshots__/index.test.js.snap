// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<List /> Should render and match the snapshot 1`] = `
<div
  class="List__SList-sc-143fwte-0 fWFyFl"
>
  <table>
    <thead>
      <tr>
        <th>
          Name
        </th>
        <th>
          is Delicious 
        </th>
        <th>
          is Healthy
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        aria-labelledby="label-e5d9d9f5"
        class="ListItem__SListItem-y53y74-0 cbkTsJ"
        role="group"
      >
        <td>
          <h5
            id="label-e5d9d9f5"
          >
            ice cream
          </h5>
        </td>
        <td>
          <label
            class="CheckBox__HiddenLabel-sc-3ds2-0 gnQLZV"
          >
            <div
              class="CheckBox__CheckboxContainer-sc-3ds2-1 hgbZRk"
            >
              <input
                checked=""
                class="CheckBox__HiddenCheckbox-sc-3ds2-3 chOmlZ"
                data-name="isDelicious"
                type="checkbox"
              />
              <div
                class="CheckBox__StyledCheckbox-sc-3ds2-4 gjKREU"
              >
                <svg
                  class="CheckBox__Icon-sc-3ds2-2 dpDFqD"
                  viewBox="0 0 24 24"
                >
                  <polyline
                    points="20 6 9 17 4 12"
                  />
                </svg>
              </div>
            </div>
            <span>
              is Delicious
            </span>
          </label>
        </td>
        <td>
          <label
            class="CheckBox__HiddenLabel-sc-3ds2-0 gnQLZV"
          >
            <div
              class="CheckBox__CheckboxContainer-sc-3ds2-1 hgbZRk"
            >
              <input
                class="CheckBox__HiddenCheckbox-sc-3ds2-3 chOmlZ"
                data-name="isHealthy"
                type="checkbox"
              />
              <div
                class="CheckBox__StyledCheckbox-sc-3ds2-4 eVKOMr"
              >
                <svg
                  class="CheckBox__Icon-sc-3ds2-2 dpDFqD"
                  viewBox="0 0 24 24"
                >
                  <polyline
                    points="20 6 9 17 4 12"
                  />
                </svg>
              </div>
            </div>
            <span>
              is Healthy
            </span>
          </label>
        </td>
      </tr>
      <tr
        aria-labelledby="label-a9ba692b"
        class="ListItem__SListItem-y53y74-0 cbkTsJ"
        role="group"
      >
        <td>
          <h5
            id="label-a9ba692b"
          >
            pizza
          </h5>
        </td>
        <td>
          <label
            class="CheckBox__HiddenLabel-sc-3ds2-0 gnQLZV"
          >
            <div
              class="CheckBox__CheckboxContainer-sc-3ds2-1 hgbZRk"
            >
              <input
                checked=""
                class="CheckBox__HiddenCheckbox-sc-3ds2-3 chOmlZ"
                data-name="isDelicious"
                type="checkbox"
              />
              <div
                class="CheckBox__StyledCheckbox-sc-3ds2-4 gjKREU"
              >
                <svg
                  class="CheckBox__Icon-sc-3ds2-2 dpDFqD"
                  viewBox="0 0 24 24"
                >
                  <polyline
                    points="20 6 9 17 4 12"
                  />
                </svg>
              </div>
            </div>
            <span>
              is Delicious
            </span>
          </label>
        </td>
        <td>
          <label
            class="CheckBox__HiddenLabel-sc-3ds2-0 gnQLZV"
          >
            <div
              class="CheckBox__CheckboxContainer-sc-3ds2-1 hgbZRk"
            >
              <input
                class="CheckBox__HiddenCheckbox-sc-3ds2-3 chOmlZ"
                data-name="isHealthy"
                type="checkbox"
              />
              <div
                class="CheckBox__StyledCheckbox-sc-3ds2-4 eVKOMr"
              >
                <svg
                  class="CheckBox__Icon-sc-3ds2-2 dpDFqD"
                  viewBox="0 0 24 24"
                >
                  <polyline
                    points="20 6 9 17 4 12"
                  />
                </svg>
              </div>
            </div>
            <span>
              is Healthy
            </span>
          </label>
        </td>
      </tr>
      <tr
        aria-labelledby="label-ze128a47"
        class="ListItem__SListItem-y53y74-0 llVrwU"
        role="group"
      >
        <td>
          <h5
            id="label-ze128a47"
          >
            spinach
          </h5>
        </td>
        <td>
          <label
            class="CheckBox__HiddenLabel-sc-3ds2-0 gnQLZV"
          >
            <div
              class="CheckBox__CheckboxContainer-sc-3ds2-1 hgbZRk"
            >
              <input
                class="CheckBox__HiddenCheckbox-sc-3ds2-3 chOmlZ"
                data-name="isDelicious"
                type="checkbox"
              />
              <div
                class="CheckBox__StyledCheckbox-sc-3ds2-4 eVKOMr"
              >
                <svg
                  class="CheckBox__Icon-sc-3ds2-2 dpDFqD"
                  viewBox="0 0 24 24"
                >
                  <polyline
                    points="20 6 9 17 4 12"
                  />
                </svg>
              </div>
            </div>
            <span>
              is Delicious
            </span>
          </label>
        </td>
        <td>
          <label
            class="CheckBox__HiddenLabel-sc-3ds2-0 gnQLZV"
          >
            <div
              class="CheckBox__CheckboxContainer-sc-3ds2-1 hgbZRk"
            >
              <input
                class="CheckBox__HiddenCheckbox-sc-3ds2-3 chOmlZ"
                data-name="isHealthy"
                type="checkbox"
              />
              <div
                class="CheckBox__StyledCheckbox-sc-3ds2-4 eVKOMr"
              >
                <svg
                  class="CheckBox__Icon-sc-3ds2-2 dpDFqD"
                  viewBox="0 0 24 24"
                >
                  <polyline
                    points="20 6 9 17 4 12"
                  />
                </svg>
              </div>
            </div>
            <span>
              is Healthy
            </span>
          </label>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
